### PR TITLE
list vm instead of get when getting virtual machine

### PR DESCRIPTION
**Pull request by zqingqing1**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
We have switched VMSS and VMSSAPI caches to LIST. For VM caches, we should also switch to LIST so that number of requests would be reduced.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: revert this change.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap_test.go [hunk 2]
```diff
@@ -7,7 +7,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap_test.go]: revert this

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go [hunk 3]
```diff
@@ -44,6 +45,8 @@ var (
 	azureResourceGroupNameRE = regexp.MustCompile(`.*/subscriptions/(?:.*)/resourceGroups/(.+)/providers/(?:.*)`)
 )
 
+const vmKey = "vmKey"
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: better be
```
const vmListCacheKey = "vmListCacheKey"
```

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go [hunk 4]
```diff
@@ -62,18 +65,44 @@ func checkResourceExistsFromError(err *retry.Error) (bool, *retry.Error) {
...
 	}
 
-	if cachedVM == nil {
-		return vm, cloudprovider.InstanceNotFound
+	if vm != nil {
+		return *vm, nil
 	}
 
-	return *(cachedVM.(*compute.VirtualMachine)), nil
+	klog.V(3).Infof("Couldn't find VM with name %s, refreshing the cache", vmName)
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: V(2)

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap_test.go (hunk 2 above)
Comment (andyzhangx) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap_test.go]: this lead to pull-kubernetes-verify error
```
Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap_test.go
```

**Review by feiskyer — COMMENTED** [review]
Thanks for the PR. Could you also correct the commit message? it should be `List virtual machines` instead of `List vmss`.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go (hunk 1 above)
Comment (zqingqing1) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: sure, not sure why that changed, but will remove it

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go (hunk 3 above)
Comment (zqingqing1) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: sounds good to me

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go (hunk 4 above)
Comment (zqingqing1) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: Maybe I asked a stupid question, how to decide the level of logging?

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by andyzhangx** [comment]
I could see below error is caused by this PR:
```
I0104 04:03:12.715] Jan  4 04:03:12.713: INFO: At 2020-01-04 03:45:34 +0000 UTC - event for azuredisk-volume-tester-hf4n4: {attachdetach-controller } FailedAttachVolume: AttachVolume.Attach failed for volume "pvc-8368035b-a1d6-4e7c-af27-fa4d0d2df178" : Retriable: false, RetryAfter: 0001-01-01 00:00:00 +0000 UTC, HTTPStatusCode: 400, RawError: compute.VirtualMachinesClient#Update: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameterConflictingProperties" Message="The property 'dataDisk.createOption' cannot have the value 'Empty' when property 'dataDisk.managedDisk.id' is set."
```
original error logs in this PR:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/86817/pull-kubernetes-e2e-azure-disk/1213297144086138881/build-log.txt

Let's revert this PR.

cc @feiskyer @zqingqing1 

thanks @chewong for the finding!
For every azure related change, we need to run all azure e2e tests

**Comment by andyzhangx** [comment]
cc @ritazh @craiglpeters  
this issue showed that our CI test is working perfectly, all azure could provider work should rely on these e2e CI tests, thanks @chewong for driving these tests!

**Comment by chewong** [comment]
> I could see below error is caused by this PR:
> 
> ```
> I0104 04:03:12.715] Jan  4 04:03:12.713: INFO: At 2020-01-04 03:45:34 +0000 UTC - event for azuredisk-volume-tester-hf4n4: {attachdetach-controller } FailedAttachVolume: AttachVolume.Attach failed for volume "pvc-8368035b-a1d6-4e7c-af27-fa4d0d2df178" : Retriable: false, RetryAfter: 0001-01-01 00:00:00 +0000 UTC, HTTPStatusCode: 400, RawError: compute.VirtualMachinesClient#Update: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameterConflictingProperties" Message="The property 'dataDisk.createOption' cannot have the value 'Empty' when property 'dataDisk.managedDisk.id' is set."
> ```
> 
> original error logs in this PR:
> https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/86817/pull-kubernetes-e2e-azure-disk/1213297144086138881/build-log.txt
> 
> Let's revert this PR.
> 
> cc @feiskyer @zqingqing1
> 
> thanks @chewong for the finding!
> For every azure related change, we need to run all azure e2e tests

I opened https://github.com/kubernetes/test-infra/pull/15899, which should run all Azure-related presubmit jobs automatically for now on so we don't have to comment `/test pull-kubernetes-e2e-...` every time. The reason why we didn't catch the disk error in this PR is that we didn't run Azure-related presubmit jobs after the author pushed a new commit - optional tests aren't re-run unless they are manually triggered.
